llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp

```diff
@@ -1207,6 +1207,49 @@ Instruction *InstCombinerImpl::commonIDivTransforms(BinaryOperator &I) {
     }
   }
 
+  // (X * Y) / (X * Z) --> Y / Z (and commuted variants)
+  if (match(Op0, m_Mul(m_Value(X), m_Value(Y)))) {
+    auto OB0HasNSW = cast<OverflowingBinaryOperator>(Op0)->hasNoSignedWrap();
+    auto OB0HasNUW = cast<OverflowingBinaryOperator>(Op0)->hasNoUnsignedWrap();
+
+    auto CreateDivOrNull = [&](Value *A, Value *B) -> Instruction * {
+      auto OB1HasNSW = cast<OverflowingBinaryOperator>(Op1)->hasNoSignedWrap();
+      auto OB1HasNUW =
+          cast<OverflowingBinaryOperator>(Op1)->hasNoUnsignedWrap();
+      const APInt *C1, *C2;
+      if (IsSigned && OB0HasNSW) {
+        if (OB1HasNSW && match(B, m_APInt(C1)) && !C1->isAllOnes())
+          return BinaryOperator::CreateSDiv(A, B);
+        if (match(A, m_APInt(C1)) && match(B, m_APInt(C2)) &&
+            !C2->isSignMask()) {
+          APInt Q, R;
+          APInt::sdivrem(*C1, *C2, Q, R);
+          if (R.isZero())
+            return replaceInstUsesWith(I, ConstantInt::get(A->getType(), Q));
+        }
+      }
+      if (!IsSigned && OB0HasNUW) {
+        if (OB1HasNUW)
+          return BinaryOperator::CreateUDiv(A, B);
+        if (match(A, m_APInt(C1)) && match(B, m_APInt(C2))) {
+          APInt Q, R;
+          APInt::udivrem(*C1, *C2, Q, R);
+          if (R.isZero())
```

Comment:
Wouldn't it be sufficient to check that `C2 <= C1` here? https://alive2.llvm.org/ce/z/B8cFaS

(Sorry if this was already discussed, I didn't follow the review.)

Reply:
You are right.